### ISSUE TITLE
builds seem to cumulate in updates-pending tag, an untagging bug?

### ISSUE THREAD

**Issue by kparal** [fedora-infra/bodhi#3944]
We have extremely high numbers of builds in `updates-pending` tags. For example:
```
$ koji list-tagged f31-updates-pending | wc -l
1493
$ koji list-tagged f32-updates-pending | wc -l
6894
```
The F32 numbers could be caused by F32 freeze, even though the number seems excessively high even when frozen. But F31 numbers are definitely wrong, because the last F31 updates push was **today**, and yet there are 1500 builds in there(!). Some builds are even duplicated/triplicated:
```
$ koji list-tagged f31-updates-pending | grep -E '^(Cython|anaconda|binutils|buildah|exiv2|hgview)'
Cython-0.29.13-1.fc31                     f31-updates-pending   ignatenkobrain
Cython-0.29.13-2.fc31                     f31-updates-pending   limb
anaconda-31.21-1.fc31                     f31-updates-pending   jkonecny
anaconda-31.22-1.fc31                     f31-updates-pending   m4rtink
binutils-2.32-22.fc31                     f31-updates-pending   nickc
binutils-2.32-23.fc31                     f31-updates-pending   nickc
binutils-2.32-24.fc31                     f31-updates-pending   nickc
buildah-1.11.0-0.11.dev.git60d5480.fc31   f31-updates-pending   lsm5
buildah-1.11.0-0.15.dev.git2df08f0.fc31   f31-updates-pending   lsm5
buildah-1.11.0-0.17.dev.git3f5436f.fc31   f31-updates-pending   lsm5
exiv2-0.27.2-0.2.RC2.fc31                 f31-updates-pending   rdieter
exiv2-0.27.2-1.fc31                       f31-updates-pending   rdieter
hgview-1.10.5-1.fc31                      f31-updates-pending   kiilerix
hgview-1.10.5-2.fc31                      f31-updates-pending   kiilerix
hgview-1.10.5-4.fc31                      f31-updates-pending   kiilerix
```
It looks like there's some bug regarding untagging, and the builds cumulate in this tag. I found this problem because our Taskotron's rpmdelint task downloads the whole tag contents, and our test machines started running out of disk space. I'll have to disable rpmdeplint until this is resolved somehow (cc @tflink).

**Comment by mattiaverga** [comment]
I've noticed that if an update is still in pending->testing and gets enough karma to reach the stable state, it will go to pending->stable and `testing-pending` tag is not removed from its builds.

For example:
https://bodhi.fedoraproject.org/updates/FEDORA-2020-4fb1b8537d
https://koji.fedoraproject.org/koji/buildinfo?buildID=1471984

I think we should address this in the composer.

**Comment by AnchalJain** [comment]
@cverna  can I do this, issue?

**Comment by cverna** [comment]
> @cverna can I do this, issue?

Sure I have not look exactly at what need to be done, but you can start to investigate :)

**Comment by mattiaverga** [comment]
Instead of address this in the composer, I think it could be better to remove the `pending_testing_tag` when we request stable in https://github.com/fedora-infra/bodhi/blob/894b795bee6f9f79fd61596db9a9586750afeb8e/bodhi/server/models.py#L2863-L2868 if the update hasn't been pushed to testing yet.

**Comment by karmadolkar** [comment]
I had expressed interest in working on this a few days ago on the IRC DM. :)
This is the patch I came up with:

```
 # Add the appropriate 'pending' koji tag to this update, so tools like 
 # AutoQA can compose repositories of them for testing. 
 if action is UpdateRequest.testing: 
     self.add_tag(self.release.pending_signing_tag) 
 elif action is UpdateRequest.stable: 
     self.add_tag(self.release.pending_stable_tag
     self.remove_tag(self.release.pending_testing_tag)
```
Is this correct?

**Comment by mattiaverga** [comment]
@karmadolkar correct, just use the `remove_tag` only when appropriate (for example we can check if `self.request` is still  `UpdateRequest.testing` at this point)

**Comment by karmadolkar** [comment]
```
 # Add the appropriate 'pending' koji tag to this update, so tools like 
 # AutoQA can compose repositories of them for testing. 
 if action is UpdateRequest.testing: 
     self.add_tag(self.release.pending_signing_tag) 
 elif action is UpdateRequest.stable: 
     self.add_tag(self.release.pending_stable_tag
     if self.request is UpdateRequest.testing:
         self.remove_tag(self.release.pending_testing_tag)
```

Is this correct?

**Comment by mattiaverga** [comment]
> 
> Is this correct?

It seems so in my opinion.

**Comment by kparal** [comment]
I'm not sure if the fix was already deployed, but the number of updates in updates-pending are still extremely high (almost unchanged).

**Comment by mattiaverga** [comment]
> I'm not sure if the fix was already deployed, but the number of updates in updates-pending are still extremely high (almost unchanged).

No, the PR has not been included in Bodhi 5.2.x
I think it's better to remove the 5.2 milestone to avoid confusion.
